src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs

```diff
@@ -0,0 +1,149 @@
+
+        private static readonly DiagnosticDescriptor DiagnosticDescriptor = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableTryGetValueMessage,
+            "Performance",
+            RuleLevel.IdeSuggestion,
+            s_localizableTryGetValueDescription,
+            false,
+            false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(DiagnosticDescriptor);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!TryGetDictionaryTypeAndMembers(context.Compilation, out var iDictionaryType, out var containsKeySymbol, out var addSymbol))
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(ctx => OnOperationAnalysis(iDictionaryType, containsKeySymbol, addSymbol, ctx), OperationKind.Conditional);
+        }
+
+        private static void OnOperationAnalysis(INamedTypeSymbol iDictionaryType, IMethodSymbol containsKeySymbol, IMethodSymbol addSymbol, OperationAnalysisContext context)
+        {
+            var conditional = (IConditionalOperation)context.Operation;
+            if (conditional.Condition is IInvocationOperation invocation
+                && IsDictionaryType(invocation.TargetMethod.ContainingType, iDictionaryType)
+                && DoesSignatureMatch(invocation.TargetMethod, containsKeySymbol)
+                && TryGetEligibleDictionaryAddLocation(conditional.WhenFalse, invocation.Instance, invocation.Arguments[0].Value, addSymbol, out var dictionaryAddLocation))
+            {
+                var additionalLocations = ImmutableArray.Create(dictionaryAddLocation);
+                context.ReportDiagnostic(Diagnostic.Create(DiagnosticDescriptor, invocation.Syntax.GetLocation(), additionalLocations));
+            }
+            else if (conditional.Condition is IUnaryOperation { OperatorKind: UnaryOperatorKind.Not, Operand: IInvocationOperation i }
+                     && IsDictionaryType(i.TargetMethod.ContainingType, iDictionaryType)
+                     && DoesSignatureMatch(i.TargetMethod, containsKeySymbol)
+                     && TryGetEligibleDictionaryAddLocation(conditional.WhenTrue, i.Instance, i.Arguments[0].Value, addSymbol, out dictionaryAddLocation))
+            {
+                var additionalLocations = ImmutableArray.Create(dictionaryAddLocation);
+                context.ReportDiagnostic(Diagnostic.Create(DiagnosticDescriptor, i.Syntax.GetLocation(), additionalLocations));
+            }
```

Comment:
A little bit of refactoring to avoid duplication.

```suggestion
            var conditional = (IConditionalOperation)context.Operation;
            var condition = conditional.Condition;
            var whenBranch = conditional.WhenFalse;
            if (condition is IUnaryOperation { OperatorKind: UnaryOperatorKind.Not } unaryOperation)
            {
                condition = unaryOperation.Operand;
                whenBranch = conditional.WhenTrue;
            }

            if (condition is IInvocationOperation invocation
                && IsDictionaryType(invocation.TargetMethod.ContainingType, iDictionaryType)
                && DoesSignatureMatch(invocation.TargetMethod, containsKeySymbol)
                && TryGetEligibleDictionaryAddLocation(conditional.WhenFalse, invocation.Instance, invocation.Arguments[0].Value, addSymbol, out var dictionaryAddLocation))
            {
                var additionalLocations = ImmutableArray.Create(dictionaryAddLocation);
                context.ReportDiagnostic(Diagnostic.Create(DiagnosticDescriptor, invocation.Syntax.GetLocation(), additionalLocations));
            }
```